_posts/memory-creation-overflow-bug.md

```diff
@@ -0,0 +1,87 @@
+---
+layout: post
+title:  "Solidity Memory Array Creation Overflow Bug"
+date:   2020-04-02
+author: Solidity Team
+---
+
+# Solidity Memory Array Creation Overflow Bug Announcement
+
+On the 28th of March a bug in the Solidity code generator was reported through the
+<a href="https://bounty.ethereum.org/" target="_blank">Ethereum Foundation Bounty program</a>,
+by John Toman of Certora. The bug is fixed with version 0.6.5, released on [[[Insert Release Date]]].
+
+## Who should be concerned
+
+If you have deployed contracts which allocates a memory array of user-supplied length,
+but does not copy or iterate over it, this can lead to memory corruption.
+In particular, a sufficiently large length during the allocation of the memory array
+will result in subsequent memory allocations to overlap with the memory region of the array.
+However, the length of the array is correctly stored, so copying or iterating over the array
+will terminate and revert the transation with out-of-gas in these cases.
+
+## How to check if contract is vulnerable
+
+If you allocate dynamic memory arrays using ``new T[](length)`` with a length that depends
+on user input, you may be vulnerable to this bug. If it is possible to supply a sufficiently
+large length, subsequent memory allocations will have overlapping memory regions and operations
+that use memory scratch space can invalidate the contents of the array.
+If you traverse or copy the array after creating it, however,
+the transaction will safely revert with out-of-gas, since the array length is stored correctly.
+
+Safe example:
+
+    uint[] memory x = new uint[](length);
+    for (uint i = 0; i < length; ++i) {
+        x[i] = i**2;
+    }
+    ...
+
+Another safe example:
+
+    contract C {
+        uint[] x;
+        function f(uint length) public {
+            x = new uint[](length);
```

Comment:
```suggestion
            // this is safe because the copy from memory to storage will
            // iterate over the whole array and go out-of-gas
            x = new uint[](length);
```